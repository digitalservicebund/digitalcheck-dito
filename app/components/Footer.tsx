import { Link } from "react-router";
import { openInNewIconElement } from "~/components/openInNewWindow";
import { footer } from "~/resources/content/components/footer";
import tailwindMerge from "~/utils/tailwindMerge";
import { A11Y_MESSAGE_NEW_WINDOW } from "./Aria";
import Image from "./Image";

type LinkProps = {
  preText?: string;
  url: string;
  text: string;
  openInNewTab?: boolean;
};

const FooterLink = ({ link }: { link: LinkProps }) => (
  <>
    {link?.preText}{" "}
    <Link
      to={link.url}
      className="text-link increase-tap-area"
      target={link.openInNewTab ? "_blank" : undefined}
      rel={link.openInNewTab ? "noreferrer" : undefined}
      aria-describedby={link.openInNewTab ? A11Y_MESSAGE_NEW_WINDOW : undefined}
    >
      {link.text} {link.openInNewTab && openInNewIconElement}
    </Link>
  </>
);

const LinkList = ({
  header,
  links,
  className,
}: {
  header?: string;
  links: (LinkProps | LinkProps[])[];
  className?: string;
}) => (
  <div className={tailwindMerge("ds-stack ds-stack-8", className)}>
    {header && <h2 className="ds-label-section">{header}</h2>}
    <ul className="list-unstyled flex flex-col gap-8">
      {links.map((link) => {
        const subLinks = Array.isArray(link) ? link : [link];
        return (
          <li key={subLinks[0].url}>
            {subLinks.map((subLink) => (
              <FooterLink key={subLink.url} link={subLink} />
            ))}
          </li>
        );
      })}
    </ul>
  </div>
);

export default function Footer() {
  return (
    <footer
      className="ds-label-03-reg flex w-full flex-row justify-center px-16 leading-snug text-gray-900 print:hidden"
      aria-label={footer.navLabel}
      id="page-footer"
    >
      <div className="ds-stack ds-stack-32 sm:ds-stack-40 w-full max-w-[1120px] border-t-2 border-blue-300 py-40">
        <nav
          className="grid grid-cols-1 justify-between gap-y-32 sm:grid-cols-[repeat(3,_minmax(0,_18rem))] sm:grid-rows-2"
          aria-label={footer.top.navLabel}
          id="quick-overview-nav"
        >
          <LinkList
            header={footer.top.supportOffer.title}
            links={footer.top.supportOffer.links}
            className="row-span-2"
          />
          <LinkList
            header={footer.top.stepByStep.title}
            links={footer.top.stepByStep.links}
            className="row-span-2"
          />
          <LinkList
            header={footer.top.basics.title}
            links={footer.top.basics.links}
          />
          <LinkList
            header={footer.top.examples.title}
            links={footer.top.examples.links}
          />
        </nav>

        <nav aria-label={footer.middle.navLabel} id="sitemap-nav">
          <ul className="list-unstyled flex flex-row flex-wrap gap-8">
            {footer.middle.links.map((link) => (
              <li
                key={link.url}
                className="after:ml-8 after:inline-block after:content-['•'] last:after:content-['']"
              >
                <FooterLink link={link} />
              </li>
            ))}
          </ul>
        </nav>

        <hr className="mt-16 w-full border-t-2 border-blue-300" />

        <div className="space-y-16 sm:flex sm:gap-48">
          <Image
            url="/logo/bmi-logo.png"
            width={120}
            alternativeText="Logo des Bundesministerium des Innern und für Heimat"
          />
          <nav aria-label={footer.bottom.navLabel} id="external-links-nav">
            <LinkList links={footer.bottom.links} />
          </nav>
        </div>
      </div>
    </footer>
  );
}
