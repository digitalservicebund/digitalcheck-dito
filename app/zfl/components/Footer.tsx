import Image from "app/components/Image";
import { Link } from "react-router";
import { openInNewIconElement } from "~/components/openInNewWindow";
import { A11Y_MESSAGE_NEW_WINDOW } from "~/resources/constants";
import tailwindMerge from "~/utils/tailwindMerge";
import { ZFL_EMAIL, ZFL_PHONE } from "../constants";
import {
  ROUTE_ZFL_A11Y,
  ROUTE_ZFL_IMPRINT,
  ROUTE_ZFL_INVOLVEMENTS,
  ROUTE_ZFL_PRIVACY,
  ROUTE_ZFL_TRAININGS,
} from "../routes";

type LinkProps = {
  preText?: string;
  postText?: string;
  url: string;
  text: string;
  openInNewTab?: boolean;
};

const FooterLink = ({ link }: { link: LinkProps }) => (
  <>
    {link?.preText}{" "}
    <Link
      to={link.url}
      className="text-link increase-tap-area"
      target={link.openInNewTab ? "_blank" : undefined}
      rel={link.openInNewTab ? "noreferrer" : undefined}
      aria-describedby={link.openInNewTab ? A11Y_MESSAGE_NEW_WINDOW : undefined}
    >
      <span className="flex items-center">
        {link.text} {link.openInNewTab && openInNewIconElement}
      </span>
    </Link>
    {link?.postText && <> {link.postText}</>}
  </>
);

const LinkList = ({
  header,
  links,
  className,
}: {
  header?: string;
  links: (LinkProps | LinkProps[])[];
  className?: string;
}) => (
  <div className={tailwindMerge("ds-stack ds-stack-8", className)}>
    {header && <h2 className="ds-label-section">{header}</h2>}
    <ul className="list-unstyled flex flex-col gap-8">
      {links.map((link) => {
        const subLinks = Array.isArray(link) ? link : [link];
        return (
          <li key={subLinks[0].url}>
            {subLinks.map((subLink) => (
              <FooterLink key={subLink.url} link={subLink} />
            ))}
          </li>
        );
      })}
    </ul>
  </div>
);

export default function Footer() {
  return (
    <footer
      id="footer"
      className="ds-label-03-reg flex w-full flex-row justify-center border-t-2 border-blue-300 leading-snug text-gray-900 print:hidden"
      aria-label="Seitenfußbereich"
    >
      <div className="w-full max-w-6xl space-y-32 px-16 py-40 sm:space-y-40">
        <nav
          className="grid grid-cols-1 items-end justify-between gap-32 sm:grid-cols-[repeat(3,_minmax(0,_18rem))] sm:grid-rows-2"
          aria-label="Schnellübersicht"
        >
          <LinkList
            header="Unterstützungsangebote"
            links={[
              [
                {
                  preText: "Support:",
                  text: ZFL_PHONE.display,
                  url: ZFL_PHONE.url,
                },
                {
                  preText: " oder",
                  text: ZFL_EMAIL.display,
                  url: ZFL_EMAIL.url,
                },
              ],
              {
                text: ROUTE_ZFL_INVOLVEMENTS.title,
                url: ROUTE_ZFL_INVOLVEMENTS.href,
              },
              {
                text: ROUTE_ZFL_TRAININGS.title,
                url: ROUTE_ZFL_TRAININGS.href,
              },
            ]}
            className="row-span-2"
          />
          <LinkList
            header=""
            links={[
              {
                text: ROUTE_ZFL_IMPRINT.title,
                url: ROUTE_ZFL_IMPRINT.href,
              },
              {
                text: ROUTE_ZFL_PRIVACY.title,
                url: ROUTE_ZFL_PRIVACY.href,
              },
              {
                text: ROUTE_ZFL_A11Y.title,
                url: ROUTE_ZFL_A11Y.href,
              },
            ]}
            className="row-span-2"
          />
        </nav>

        <hr className="mt-16 w-full border-t-2 border-blue-300" />

        <nav aria-label="Externe Verlinkungen" className="space-y-40">
          <div className="space-y-16 sm:flex sm:flex-row sm:items-center sm:gap-48 sm:space-y-0">
            <Image
              url="/logo/bmds-logo.png"
              width={120}
              alternativeText="Logo des Bundesministerium für Digitales und Staatsmodernisierung"
            />

            <div className="max-w-2xl space-y-8">
              <h3 className="ds-label-03-bold">Federführung</h3>
              <LinkList
                links={[
                  {
                    preText:
                      "Das Zentrum für Legistik entsteht im Auftrag des ",
                    text: "Bundesministerium für Digitales und Staatsmodernisierung",
                    url: "https://bmds.bund.de/",
                  },
                ]}
              />
            </div>
          </div>
        </nav>
      </div>
    </footer>
  );
}
