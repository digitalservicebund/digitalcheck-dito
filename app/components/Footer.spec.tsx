import { render } from "@testing-library/react";
import { createRoutesStub } from "react-router";
import { describe, expect, it } from "vitest";
import Footer from "./Footer";

const RouterStubFooter = createRoutesStub([
  {
    path: "/",
    Component: () => Footer(),
  },
]);

describe("Footer Component", () => {
  it("Renders correctly", () => {
    const { container } = render(<RouterStubFooter />);

    expect(container).toMatchSnapshot();
  });
});
