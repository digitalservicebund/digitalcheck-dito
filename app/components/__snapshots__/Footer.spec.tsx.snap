// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Footer Component > Renders correctly 1`] = `
<div>
  <footer
    aria-label="Seitenfußbereich"
    class="ds-label-03-reg flex w-full flex-row justify-center px-16 leading-snug text-gray-900 print:hidden"
  >
    <div
      class="ds-stack ds-stack-32 sm:ds-stack-40 w-full max-w-[1120px] border-t-2 border-blue-300 py-40"
    >
      <nav
        aria-label="Schnellübersicht"
        class="grid grid-cols-1 justify-between gap-y-32 sm:grid-cols-[repeat(3,_minmax(0,_18rem))] sm:grid-rows-2"
      >
        <div
          class="ds-stack ds-stack-8 row-span-2"
        >
          <h2
            class="ds-label-section"
          >
            Unterstützungsangebote Für bund, Länder, Kommunen und EU-Staaten
          </h2>
          <ul
            class="list-unstyled flex flex-col gap-8"
          >
            <li>
              Support:
               
              <a
                class="text-link increase-tap-area"
                href="tel:015140767839"
              >
                0151/40 76 78 39
                 
              </a>
              oder
               
              <a
                class="text-link increase-tap-area"
                href="mailto:digitalcheck@digitalservice.bund.de"
              >
                digitalcheck@digitalservice.bund.de
                 
              </a>
            </li>
            <li>
               
              <a
                class="text-link increase-tap-area"
                data-discover="true"
                href="/unterstuetzung"
              >
                Alle Unterstützungsangebote
                 
              </a>
            </li>
            <li>
               
              <a
                class="text-link increase-tap-area"
                data-discover="true"
                href="/unterstuetzung#angebote"
              >
                Schulungen
                 
              </a>
            </li>
            <li>
               
              <a
                class="text-link increase-tap-area"
                data-discover="true"
                href="/interoperabel/nationale-kontaktstelle"
              >
                Nationale Kontaktstelle für ein interoperables Europa (2024/903 Art. 17)
                 
              </a>
            </li>
          </ul>
        </div>
        <div
          class="ds-stack ds-stack-8 row-span-2"
        >
          <h2
            class="ds-label-section"
          >
            Schritt für Schritt
          </h2>
          <ul
            class="list-unstyled flex flex-col gap-8"
          >
            <li>
               
              <a
                class="text-link increase-tap-area"
                data-discover="true"
                href="/vorpruefung"
              >
                1. Vorprüfung: Digitalbezug einschätzen
                 
              </a>
            </li>
            <li>
               
              <a
                class="text-link increase-tap-area"
                data-discover="true"
                href="/methoden"
              >
                2. Digitaltauglichkeit der Regelung sicherstellen
                 
              </a>
            </li>
            <li>
               
              <a
                class="text-link increase-tap-area"
                data-discover="true"
                href="/dokumentation"
              >
                3. Dokumentieren der Digitaltauglichkeit
                 
              </a>
            </li>
          </ul>
        </div>
        <div
          class="ds-stack ds-stack-8"
        >
          <h2
            class="ds-label-section"
          >
            Grundlagen
          </h2>
          <ul
            class="list-unstyled flex flex-col gap-8"
          >
            <li>
               
              <a
                class="text-link increase-tap-area"
                data-discover="true"
                href="/grundlagen/fuenf-prinzipien"
              >
                Prinzipien der Digitaltauglichkeit
                 
              </a>
            </li>
            <li>
               
              <a
                class="text-link increase-tap-area"
                data-discover="true"
                href="/grundlagen/methoden"
              >
                Übersicht der Werkzeuge und Methoden
                 
              </a>
            </li>
            <li>
               
              <a
                class="text-link increase-tap-area"
                data-discover="true"
                href="/interoperabel"
              >
                EU-Interoperabilität
                 
              </a>
            </li>
          </ul>
        </div>
        <div
          class="ds-stack ds-stack-8"
        >
          <h2
            class="ds-label-section"
          >
            Beispiele
          </h2>
          <ul
            class="list-unstyled flex flex-col gap-8"
          >
            <li>
               
              <a
                class="text-link increase-tap-area"
                data-discover="true"
                href="/beispiele/prinzipien"
              >
                Digitaltauglichkeit im Regelungstext
                 
              </a>
            </li>
            <li>
               
              <a
                class="text-link increase-tap-area"
                data-discover="true"
                href="/beispiele/visualisierungen"
              >
                Visualisierungen
                 
              </a>
            </li>
          </ul>
        </div>
      </nav>
      <nav
        aria-label="Sitemap"
      >
        <ul
          class="list-unstyled flex flex-row flex-wrap gap-8"
        >
          <li
            class="after:ml-8 after:inline-block after:content-['•'] last:after:content-['']"
          >
             
            <a
              class="text-link increase-tap-area"
              data-discover="true"
              href="/impressum"
            >
              Impressum
               
            </a>
          </li>
          <li
            class="after:ml-8 after:inline-block after:content-['•'] last:after:content-['']"
          >
             
            <a
              class="text-link increase-tap-area"
              data-discover="true"
              href="/datenschutz"
            >
              Datenschutzerklärung
               
            </a>
          </li>
          <li
            class="after:ml-8 after:inline-block after:content-['•'] last:after:content-['']"
          >
             
            <a
              class="text-link increase-tap-area"
              data-discover="true"
              href="/barrierefreiheit"
            >
              Barrierefreiheit
               
            </a>
          </li>
          <li
            class="after:ml-8 after:inline-block after:content-['•'] last:after:content-['']"
          >
             
            <a
              class="text-link increase-tap-area"
              data-discover="true"
              href="/sitemap"
            >
              Sitemap
               
            </a>
          </li>
          <li
            class="after:ml-8 after:inline-block after:content-['•'] last:after:content-['']"
          >
             
            <a
              aria-describedby="a11y-message--new-window"
              class="text-link increase-tap-area"
              href="https://github.com/digitalservicebund/digitalcheck-dito"
              rel="noreferrer"
              target="_blank"
            >
              Open Source Code
               
              <svg
                aria-hidden="true"
                class="mb-4 ml-[0.2em] inline-block! fill-current"
                data-testid="OpenInNewIcon"
                focusable="false"
                height="1.2em"
                viewBox="0 0 24 24"
                width="1.2em"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M0 0h24v24H0z"
                  fill="none"
                />
                <path
                  d="M19 19H5V5h7V3H5c-1.11 0-2 .9-2 2v14c0 1.1.89 2 2 2h14c1.1 0 2-.9 2-2v-7h-2v7zM14 3v2h3.59l-9.83 9.83 1.41 1.41L19 6.41V10h2V3h-7z"
                />
              </svg>
            </a>
          </li>
        </ul>
      </nav>
      <hr
        class="mt-16 w-full border-t-2 border-blue-300"
      />
      <div
        class="space-y-16 sm:flex sm:gap-48"
      >
        <img
          alt="Logo des Bundesministerium des Innern und für Heimat"
          src="/logo/bmi-logo.png"
          width="120"
        />
        <nav
          aria-label="Externe Verlinkungen"
        >
          <div
            class="ds-stack ds-stack-8"
          >
            <ul
              class="list-unstyled flex flex-col gap-8"
            >
              <li>
                Ein Onlinedienst der
                 
                <a
                  aria-describedby="a11y-message--new-window"
                  class="text-link increase-tap-area"
                  href="https://digitalservice.bund.de/"
                  rel="noreferrer"
                  target="_blank"
                >
                  DigitalService GmbH des Bundes
                   
                  <svg
                    aria-hidden="true"
                    class="mb-4 ml-[0.2em] inline-block! fill-current"
                    data-testid="OpenInNewIcon"
                    focusable="false"
                    height="1.2em"
                    viewBox="0 0 24 24"
                    width="1.2em"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      d="M0 0h24v24H0z"
                      fill="none"
                    />
                    <path
                      d="M19 19H5V5h7V3H5c-1.11 0-2 .9-2 2v14c0 1.1.89 2 2 2h14c1.1 0 2-.9 2-2v-7h-2v7zM14 3v2h3.59l-9.83 9.83 1.41 1.41L19 6.41V10h2V3h-7z"
                    />
                  </svg>
                </a>
              </li>
              <li>
                Im Auftrag des
                 
                <a
                  aria-describedby="a11y-message--new-window"
                  class="text-link increase-tap-area"
                  href="https://www.bmi.bund.de/"
                  rel="noreferrer"
                  target="_blank"
                >
                  Bundesministerium des Innern und für Heimat
                   
                  <svg
                    aria-hidden="true"
                    class="mb-4 ml-[0.2em] inline-block! fill-current"
                    data-testid="OpenInNewIcon"
                    focusable="false"
                    height="1.2em"
                    viewBox="0 0 24 24"
                    width="1.2em"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      d="M0 0h24v24H0z"
                      fill="none"
                    />
                    <path
                      d="M19 19H5V5h7V3H5c-1.11 0-2 .9-2 2v14c0 1.1.89 2 2 2h14c1.1 0 2-.9 2-2v-7h-2v7zM14 3v2h3.59l-9.83 9.83 1.41 1.41L19 6.41V10h2V3h-7z"
                    />
                  </svg>
                </a>
              </li>
            </ul>
          </div>
        </nav>
      </div>
    </div>
  </footer>
</div>
`;
